core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/util/ProgressListenerUtils.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.core.internal.util;
+
+import static software.amazon.awssdk.http.Header.CONTENT_LENGTH;
+
+import java.util.concurrent.atomic.AtomicLong;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.RequestOverrideConfiguration;
+import software.amazon.awssdk.core.SdkRequest;
+import software.amazon.awssdk.core.SdkResponse;
+import software.amazon.awssdk.core.internal.metrics.BytesReadTrackingInputStream;
+import software.amazon.awssdk.core.internal.metrics.BytesReadTrackingPublisher;
+import software.amazon.awssdk.core.internal.progress.listener.ProgressUpdater;
+import software.amazon.awssdk.http.AbortableInputStream;
+import software.amazon.awssdk.http.ContentStreamProvider;
+import software.amazon.awssdk.http.SdkHttpHeaders;
+import software.amazon.awssdk.http.async.SdkHttpContentPublisher;
+import software.amazon.awssdk.utils.StringUtils;
+
+@SdkInternalApi
+public final class ProgressListenerUtils {
+
+    private ProgressListenerUtils() {
+    }
+
+    public static SdkHttpContentPublisher wrapRequestProviderWithByteTrackingIfProgressListenerAttached(
+        SdkHttpContentPublisher requestProvider, ProgressUpdater progressUpdater) {
+        return new BytesReadTrackingPublisher(requestProvider, new AtomicLong(0L),
```

Comment:
Same here, `new AtomicLong(0L),` will break existing code path because the SDK relies on `SdkInternalExecutionAttribute.RESPONSE_BYTES_READ` to contain the counter

Reply:
hmm.. Im not modifying the context attribute SdkInternalExecutionAttribute.RESPONSE_BYTES_READ here. While wrapping the request provider(while sending the request) with BytesReadTracking publisher im setting bytes read to 0, as we have not yet made the request at that point. This should not modify what we set on the response